src/details/ArborX_DetailsBruteForceImpl.hpp

```diff
@@ -53,26 +53,94 @@ struct BruteForceImpl
   static void query(ExecutionSpace const &space, Primitives const &primitives,
                     Predicates const &predicates, Callback const &callback)
   {
+    using TeamPolicy = Kokkos::TeamPolicy<ExecutionSpace>;
     using AccessPrimitives = AccessTraits<Primitives, PrimitivesTag>;
     using AccessPredicates = AccessTraits<Predicates, PredicatesTag>;
+    using PredicateType = typename AccessTraitsHelper<AccessPredicates>::type;
+    using PrimitiveType = typename AccessTraitsHelper<AccessPrimitives>::type;
 
     int const n_primitives = AccessPrimitives::size(primitives);
     int const n_predicates = AccessPredicates::size(predicates);
+    int max_scratch_size = TeamPolicy::scratch_size_max(0) / 2;
+    int const predicates_per_team = max_scratch_size / sizeof(PredicateType);
+    int const primitives_per_team = max_scratch_size / sizeof(PrimitiveType);
+
+    int const n_primitive_tiles =
+        ceil((float)n_primitives / primitives_per_team);
+    int const n_predicate_tiles =
+        ceil((float)n_predicates / predicates_per_team);
+    int const n_teams = n_primitive_tiles * n_predicate_tiles;
+
+    using ScratchPredicateType =
+        Kokkos::View<PredicateType *,
+                     typename ExecutionSpace::scratch_memory_space,
+                     Kokkos::MemoryTraits<Kokkos::Unmanaged>>;
+    using ScratchPrimitiveType =
+        Kokkos::View<PrimitiveType *,
+                     typename ExecutionSpace::scratch_memory_space,
+                     Kokkos::MemoryTraits<Kokkos::Unmanaged>>;
+    int scratch_size = ScratchPredicateType::shmem_size(predicates_per_team) +
+                       ScratchPrimitiveType::shmem_size(primitives_per_team);
 
     Kokkos::parallel_for(
         "ArborX::BruteForce::query::spatial::"
         "check_all_predicates_against_all_primitives",
-        Kokkos::MDRangePolicy<ExecutionSpace, Kokkos::Rank<2>>(
-            space, {0, 0}, {n_primitives, n_predicates}),
-        KOKKOS_LAMBDA(int primitive_index, int predicate_index) {
-          auto const &predicate =
-              AccessPredicates::get(predicates, predicate_index);
-          auto const &primitive =
-              AccessPrimitives::get(primitives, primitive_index);
-          if (predicate(primitive))
+        TeamPolicy((long)n_teams, Kokkos::AUTO)
```

Comment:
It seems that at the moment the execution space `space` is not used. I assume we want to provide it somewhere in the `TeamPolicy` constructor, similarly as how we do in `RangePolicy`?